File: src/main/java/no/nav/pensjon/regler/domain/beregning2011/JusteringsInformasjon.java
package no.nav.pensjon.regler.domain.beregning2011;

import no.nav.pensjon.regler.domain.kode.JusteringsTypeCti;

import java.io.Serializable;
import java.lang.reflect.Constructor;
import java.lang.reflect.InvocationTargetException;
import java.util.ArrayList;
import java.util.List;

public class JusteringsInformasjon implements Serializable {
    private static final long serialVersionUID = 2474123061927464485L;

    private double totalJusteringsfaktor;
    private JusteringsTypeCti justeringsTypeCti;
    private List<IJustering> elementer;

    public JusteringsInformasjon() {
        super();
        elementer = new ArrayList<IJustering>();
    }

    public JusteringsInformasjon(JusteringsInformasjon ji) {
        this();
        totalJusteringsfaktor = ji.totalJusteringsfaktor;
        if (ji.justeringsTypeCti != null) {
            justeringsTypeCti = new JusteringsTypeCti(ji.getJusteringsTypeCti());
        }
        if (ji.elementer != null) {
            for (IJustering ij : ji.elementer) {
                Class<? extends IJustering> clazz = ij.getClass();
                try {
                    Constructor<?> constructor = clazz.getConstructor(new Class[] {clazz});
                    elementer.add((IJustering) constructor.newInstance(ij));
                } catch (InvocationTargetException e) {
                    //Vil kastes hvis copy constructor f.eks. ledet til nullpointerexception.
                    Throwable cause = e.getCause();
                    if (cause instanceof RuntimeException) {
                        throw (RuntimeException) cause;
                    } else {
                        throw new RuntimeException(e);
                    }
                } catch (Exception e) {
                    //Vil kastes hvis f.eks. copy constructor ikke finnes for komponenten.
                    throw new RuntimeException(e);
                }
            }
        }
    }

    public JusteringsTypeCti getJusteringsTypeCti() {
        return justeringsTypeCti;
    }

    public void setJusteringsTypeCti(JusteringsTypeCti justeringsTypeCti) {
        this.justeringsTypeCti = justeringsTypeCti;
    }

    public double getTotalJusteringsfaktor() {
        return totalJusteringsfaktor;
    }

    public void setTotalJusteringsfaktor(double totalJusteringsfaktor) {
        this.totalJusteringsfaktor = totalJusteringsfaktor;
    }

    public List<IJustering> getElementer() {
        return elementer;
    }

    public void setElementer(List<IJustering> elementer) {
        this.elementer = elementer;
    }
}

File: src/main/kotlin/no/nav/pensjon/regler/domain/beregning2011/JusteringsInformasjon.kt
package no.nav.pensjon.regler.domain.beregning2011

import no.nav.pensjon.regler.domain.enum.JusteringsTypeEnum
import java.io.Serializable
import java.lang.reflect.Constructor
import java.lang.reflect.InvocationTargetException

open class JusteringsInformasjon : Serializable {
    var totalJusteringsfaktor = 0.0
    var justeringsTypeEnum: JusteringsTypeEnum? = null

    var elementer: MutableList<IJustering> = mutableListOf()

    constructor()
    constructor(ji: JusteringsInformasjon) : this() {
        totalJusteringsfaktor = ji.totalJusteringsfaktor
        if (ji.justeringsTypeEnum != null) {
            justeringsTypeEnum = ji.justeringsTypeEnum
        }
        for (ij in ji.elementer) {
            val clazz: Class<out IJustering> = ij.javaClass
            try {
                val constructor: Constructor<*> = clazz.getConstructor(*arrayOf<Class<*>>(clazz))
                elementer.add((constructor.newInstance(ij) as IJustering))
            } catch (e: InvocationTargetException) {
                //Vil kastes hvis copy constructor f.eks. ledet til nullpointerexception.
                val cause = e.cause
                if (cause is RuntimeException) {
                    throw (cause as RuntimeException?)!!
                } else {
                    throw RuntimeException(e)
                }
            } catch (e: Exception) {
                //Vil kastes hvis f.eks. copy constructor ikke finnes for komponenten.
                throw RuntimeException(e)
            }
        }
    }

    constructor(
        totalJusteringsfaktor: Double = 0.0,
        justeringsTypeEnum: JusteringsTypeEnum? = null,
        elementer: MutableList<IJustering> = mutableListOf()
    ) {
        this.totalJusteringsfaktor = totalJusteringsfaktor
        this.justeringsTypeEnum = justeringsTypeEnum
        this.elementer = elementer
    }
}